enhancements/kata-containers/kata-containers-dev-preview.md

```diff
@@ -0,0 +1,276 @@
+3. QEMU RPMs have been tested / validated using an identical environment
+4. No additional work with regarding to re-packaging those in another form other than the RPMs which is the way they are consumed by other Layered Products
+
+**Cons:**
+1. 20 MB extra, installed in the host, for those who'd be using kata runtime
+2. Updates / Removal may be more complicated than having the RPMs as part of machine-os-content
+
+#### Long term: Use RHCOS extensions (qemu-kiwi and dependencies only)
+In this option we build upon the RHCOS extension approach planned for OCP4.6.
+
+**What is the qemu-kiwi RHCOS extension?**
+
+This will add additional software onto the host - but this software will still be versioned with the host (included as part of the OpenShift release payload) and upgraded with the cluster.
+
+The additional RPMs required for qemu-kiwi and its dependencies are:
+* qemu-kiwi
+* qemu-kvm-common
+* pxe-roms-qemu
+* libpmem
+* pixman
+* seabios-bin
+* seavgabions-bin
+* sgabios-bin
+
+See additional details about RHCOS extensions in general in: https://github.com/openshift/enhancements/pull/317
+
+**So What do we want to do with this?**
+
+We start by creating a container image with only the kata upstream components as static binaries (no RHEL-AV components). The kata operator will pull down this image and mount it in the host mount namespace, i.e. without running it as a container. We then use the RHCOS extensions for deploying the qemu-kiwi RHEL-AV RPMs.
+
+**Pros:**
+1. RHEL-AV depends on libraries and other tools which are all available as RPMs. Turning this into a static image instead of RPMs creates a huge single binary which we avoid with extensions
+2. RHEL-AV has been tested / validated as RPMs and not a single binary which means we saves the extra overhead
+3. RHEL-AV RPMs can be installed only by the customers who would be interested in using the kata runtime. QEMU (and its dependencies) get updated as part of the machine-os-content updates
+4. There's no extra work related to properly loading qemu-kiwi dependencies, as all the packages would be installed in the same PATH its build in
+5. Better control of the versions that are installed both for qemu (specific version for a given host) and for the dependencies (tracked by rpm)
+
+**Cons:**
+1. ~20 mb extra, installed in the host, for those who'd be using kata runtime
+2. machine-os-content would still carry this until there is a separate machine-os-content- extensions container
+
+
+
+### KataContainers Operator development
+The goal is to develop a kata operator that can be used to manage the entire lifecycle of kata components in an OpenShift cluster. We will have a controller that watches for a kata **custom resource (CR)** and a daemon-set that acts upon changes to the (CR) and do the required work on the worker nodes (install, uninstall update,...).
+
+Via the CR it will be possible to select a subset of worker nodes.
```

Comment:
Why is this out of curiosity?  Why not e.g. enable the feature on all available nodes but use regular `nodeSelector` for steering?

Is it because it fully takes over the crio backend and so hence all containers on that node are kata?  That seems unfortunate if true.

Reply:
Hey @cgwalters!

> Why is this out of curiosity? 
The default behaviour is to deploy on all available **worker** nodes.  However, it's been decided to leave the option for the admin to select the nodes they want, if they don't want all the worker nodes using Kata Containers.

> Why not e.g. enable the feature on all available nodes but use regular nodeSelector for steering?
That would simplify our code, indeed. Shall we consider removing the option from the operator?

> Is it because it fully takes over the crio backend and so hence all containers on that node are kata? That seems unfortunate if true.
No, that's not the case, at all.  A note capable of doing Kata Containers should not, and will not, lose its ability to do "vanilla" containers.

@cgwalters, have I addressed all your concerns?

Reply:
> The default behaviour is to deploy on all available worker nodes. 

Sure, seems reasonable.  In a compact cluster the control plane are workers too so this will apply there.

> That would simplify our code, indeed. Shall we consider removing the option from the operator?

I don't have a really strong opinion on this but if having admins steer the pods with regular selectors works then why not use that?  Less API surface for you to maintain, and feels more native to admins, etc.

Reply:
There is scheduling support for runtime classes that we can make use of. The enhancement should mention that.
https://v1-16.docs.kubernetes.io/docs/concepts/containers/runtime-class/#